libraries/chain/peer_keys_db.cpp

```diff
@@ -0,0 +1,79 @@
+#include <eosio/chain/peer_keys_db.hpp>
+#include <eosio/chain/contract_table_objects.hpp>
+
+namespace eosio::chain {
+
+peer_keys_db_t::peer_keys_db_t() {}
+
+std::optional<public_key_type> peer_keys_db_t::operator[](name n) const {
+   std::lock_guard g(_m);
+   if (auto it = _peer_key_map.find(n); it != _peer_key_map.end())
+      return std::optional<public_key_type>(it->second);
+   return std::optional<public_key_type>{};
+}
+
+size_t peer_keys_db_t::size() const {
+   std::lock_guard g(_m);
+   return _peer_key_map.size();
+}
+
+// we update the keys that were registered up to lib_number (inclusive)
+// --------------------------------------------------------------------
+size_t peer_keys_db_t::update_peer_keys(const controller& chain, uint32_t lib_number) {
+   std::lock_guard g(_m);
+   size_t num_updated = 0;
+   try {
+      if (lib_number <= _block_num)
+         return num_updated;                      // nothing to do
+
+      const auto& db = chain.db();
+      const auto  table_ref = boost::make_tuple("eosio"_n, "eosio"_n, "peerkeys"_n);
+      const auto* t_id      = db.find<table_id_object, by_code_scope_table>(table_ref);
+      EOS_ASSERT(t_id != nullptr, plugin_exception, "cannot retrieve `peerkeys` table");
```

Comment:
`plugin_exception` should be something else. Either a new one or maybe just `misc_exception`. Since this is caught and dropped below, misc is probably fine. Maybe better would be to `elog` and return instead of throw and catch below.

Reply:
Same goes for the `plugin_exception`s below. There the `misc_exception` is likely easier.

Reply:
> Maybe better would be to `elog` and return instead of throw and catch below.

How would it be better? this is not something that is expected to happen ever.